### PR TITLE
feat: add stable diffusion textual inversion example

### PR DESCRIPTION
**Pull request by garrett361**
## Description

Add textual inversion example using Hugging Face's Diffusers and Core API.

## Test Plan
Test the following:
* Do the default `finetune_const.yaml` and `generate_grid.yaml` configs give good results?
* Is it easy to run the notebook/scripts?
* Can you train your own example and then generate decent art with it?
* Are the README instructions clear?
* Does Pause/Resume work? 
* Does training/loading-then-generating with multiple concepts work?

## Checklist
- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.
- [ ] If modifying `/webui/react/src/shared/` verify `make -C webui/react test-shared` passes.

### CODE REVIEW

File: examples/diffusion/textual_inversion_stable_diffusion/detsd/trainer.py [hunk 1]
```diff
@@ -0,0 +1,690 @@
...
+        # Convert sample images to latent space.
+        prompts, img_tensors = batch
+        with torch.no_grad():
+            latent_dist = self.vae.encode(img_tensors).latent_dist
+            latents = latent_dist.sample()
+
+            # In 2112.10752, it was found that the latent space variance plays a large role in image
+            # quality.  The following scale factor helps to maintain unit latent variance.  See
+            # https://github.com/huggingface/diffusers/issues/437 for more details.
+            scale_factor = 0.18215
```
Comment (coreystatendet) [inline examples/diffusion/textual_inversion_stable_diffusion/detsd/trainer.py:282]: praise: Good documentation of magic numbers

**Review by coreystatendet — APPROVED** [review]
Overall looks great!  I know you were checking on how to deal with the Huggingface auth token for adding tests, which would be my main request if it's feasible.  Also needs a release note still.

File: examples/diffusion/textual_inversion_stable_diffusion/detsd/pipeline.py [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+        )
+
+        if isinstance(checkpoint_paths, str):
+            checkpoint_paths = [pathlib.Path(checkpoint_paths)]
+        if isinstance(checkpoint_paths, pathlib.Path):
+            checkpoint_paths = [checkpoint_paths]
+
+        for path in checkpoint_paths:
+            if isinstance(path, str):
+                path = pathlib.Path(path)
```
[the remark this answers is not in the document]
Comment (garrett361) [reply examples/diffusion/textual_inversion_stable_diffusion/detsd/pipeline.py]: Ended up refactoring this code due to other feedback, but appreciated this! Would have been cleaner.

File: examples/diffusion/textual_inversion_stable_diffusion/detsd/trainer.py [hunk 3]
```diff
@@ -0,0 +1,688 @@
...
+        num_blank_prompts: int = 0,
+        num_a_prompts: int = 0,
+        generate_training_images: bool = True,
+        inference_batch_size: int = 1,
+        num_pipeline_calls: int = 1,
+        inference_prompts: Optional[Union[str, Sequence[str]]] = None,
+        inference_scheduler_name: Literal["ddim", "lms-discrete", "pndm"] = "pndm",
+        num_inference_steps: int = 50,
+        guidance_scale: float = 7.5,
+        main_process_generator_seed: int = 2147483647,
```
[the remark this answers is not in the document]
Comment (garrett361) [reply examples/diffusion/textual_inversion_stable_diffusion/detsd/trainer.py:66]: Yes, it's the seed chosen in the `torch.Generator` [example](https://pytorch.org/docs/stable/generated/torch.Generator.html#torch.Generator.manual_seed) which follows their advice for choosing a large seed that has a good balance of 0 and 1 bits (presumably).  The implied importance of the seed choice is concerning, but that's a torch issue.

File: examples/diffusion/textual_inversion_stable_diffusion/detsd/pipeline.py [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+        # The below attrs are non-trivially instantiated as necessary through appropriate methods.
+        self.all_checkpoint_dirs = []
+        self.learned_embeddings_dict = {}
+        self.concept_to_dummy_tokens_map = {}
+        self.all_added_concepts = []
+
+        self._build_pipeline(disable_progress_bar=self.disable_progress_bar)
+
+    @classmethod
+    def generate_on_cluster(cls) -> None:
```
[the remark this answers is not in the document]
Comment (garrett361) [reply examples/diffusion/textual_inversion_stable_diffusion/detsd/pipeline.py:78]: Mmm, yeah, can definitely see that being confusing. Same issue for `train_on_cluster` in `DetSDTextualInversionTrainer`, presumably?

Both methods start with a 
```python
info = det.get_cluster_info()
assert info is not None, "z=xxx_on_cluster() must be called on a Determined cluster."
```
block, but it would be nice to not have to find that out from an assertion error.

File: examples/diffusion/textual_inversion_stable_diffusion/README.md [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+This will submit an experiment which introduces a new embedding vector into the world of Stable
+Diffusion which we will fine-tune to correspond to the concept of the Determined AI logo, as
+represented
+through training images found in `det_logos/`, such as the example found below (placed on a
+background for
+improved results):
+
+![det-logo](./det_logos/on_a_dark_blue_oil_painting_of_ocean_waves.jpg)
+
+A corresponding concept token, chosen to be `<det-logo>` as specified in the `concept_tokens` field
```
[the remark this answers is not in the document]
Comment (garrett361) [reply examples/diffusion/textual_inversion_stable_diffusion/README.md:60]: They don't, the concepts should just be strings that don't already exist in the tokenizer, which the <> ensure.

This plus some of the other issues you flagged (which have also caused others confusion) came from the [Hugging Face example](https://huggingface.co/docs/diffusers/v0.4.1/en/training/text_inversion), which I was initially adhering to much more closely.  Going to remove the <> and related issues from HF, since they have come up repeatedly, thanks!

File: examples/diffusion/textual_inversion_stable_diffusion/detsd/pipeline.py [hunk 6]
```diff
@@ -0,0 +1,351 @@
+import json
```
[the remark this answers is not in the document]
Comment (garrett361) [reply examples/diffusion/textual_inversion_stable_diffusion/detsd/pipeline.py:1]: Agreed, I think I went down that path at some point, but the base class was getting too unwieldy during that first attempt.  The other approach I considered was just subclassing the HF Pipeline class, but that also got messy due to the number of methods I would have needed to add or overwrite.

File: examples/diffusion/textual_inversion_stable_diffusion/detsd/trainer.py (hunk 1 above)
Comment (garrett361) [reply examples/diffusion/textual_inversion_stable_diffusion/detsd/trainer.py:282]: Spent a while trying to track this one down. Very annoying.

File: examples/diffusion/textual_inversion_stable_diffusion/detsd/utils.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    replacements, as well as the string representation of the dummies.
+    """
+    initializer_ids = tokenizer(
+        initializer_tokens,
+        padding="max_length",
+        truncation=True,
+        max_length=tokenizer.model_max_length,
+        return_tensors="pt",
+        add_special_tokens=False,
+    ).input_ids
```
[the remark this answers is not in the document]
Comment (garrett361) [reply examples/diffusion/textual_inversion_stable_diffusion/detsd/utils.py:24]: Agree, this was the other piece of notation imported from the [HF example](https://huggingface.co/docs/diffusers/v0.4.1/en/training/text_inversion). Will just change it, rather than stick with the original nomenclature.

File: examples/diffusion/textual_inversion_stable_diffusion/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (garrett361) [reply examples/diffusion/textual_inversion_stable_diffusion/README.md:60]: And actually, it's alright if the concepts do exist in the tokenizer, practically speaking, they'll just be replaced by dummy strings before being processed by the tokenizer anyway.

File: examples/diffusion/textual_inversion_stable_diffusion/detsd/pipeline.py [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+        - `main_process_generator_seed`: an integer specifying the seed used by the chief worker for
+        generation. Other workers add their process index to this value.
+        - `save_freq`: an integer specifying how often to write images to tensorboard.
+        - `pipeline`: containing all __init__ args
+        - `uuids`: a (possibly empty) array of any checkpoint UUIDs which are to be loaded into
+        the pipeline.
+        - `local_checkpoint_paths`: a (possibly empty) array of any checkpoint paths which are to be
+        loaded into the pipeline.
+        - `call_kwargs`: all arguments which are to be passed to the `__call__` method.
+        """
```
[the remark this answers is not in the document]
Comment (garrett361) [reply examples/diffusion/textual_inversion_stable_diffusion/detsd/pipeline.py:92]: Also agreed, wanted to make it more similar to the trainer code, but was finding it messy to do so.  Will try.

File: examples/diffusion/textual_inversion_stable_diffusion/detsd/pipeline.py [hunk 9]
```diff
@@ -0,0 +1,360 @@
...
+            if latest_checkpoint is not None:
+                with core_context.checkpoint.restore_path(latest_checkpoint) as path:
+                    with open(path.joinpath("metadata.json"), "r") as f:
+                        metadata_dict = json.load(f)
+                        generator_state_dict = torch.load(
+                            path.joinpath("generator_state_dict.pt"),
+                        )
+                        steps_completed = metadata_dict["steps_completed"]
+                        num_generated_imgs = metadata_dict["num_generated_imgs"]
+                        generator.set_state(generator_state_dict[device])
```
[the remark this answers is not in the document]
Comment (garrett361) [reply examples/diffusion/textual_inversion_stable_diffusion/detsd/pipeline.py:179]: I don't think this is true? Or at least, I think the behavior I have coded up is more desirable, or at least equally reasonable.

The two cases discussed in the link aren't about Forking, which doesn't carry over checkpoint info (just tested). The cases are instead:
1. Pause then resume in Web UI.
2. Continue from Web UI.

The advice in the second bullet point in the link is about the Continue behavior and it would give the following behavior (for example):
1. Start a Trial with `max_length = 100`.
2. After completion, hit Continue and set `max_length = 100`.
3. The Trial starts over at `steps_completed = 0` and runs for another 100 steps. 
4. After completion, you have a Trial that lists `max_length = 100` even though it was actually trained for 200 steps.

I have it coded up so that whatever you put in `max_length` after hitting Continue is the actual total `max_length` that gets put in the config. So instead of the above, w/ this code you'd have
1. Start a Trial with `max_length = 100`.
2. After completion, hit Continue and set `max_length = 100`.
3. The Trial has its `max_length` bumped up to 200 and continues until 200 total steps have been completed.

I (obviously) find my implementation more straightforward, but I suppose I should follow the docs. Or, we should change the docs.

Thoughts?

[the remark this answers is not in the document]
Comment (coreystatendet) [reply examples/diffusion/textual_inversion_stable_diffusion/detsd/pipeline.py:179]: Woops sorry, looks like I misremembered re: forking!  In that case, I agree that this seems reasonable and I'm not sure why the alternative interpretation was used in the docs.  The fact that it's open for interpretation is a bit confusing.  Maybe @rb-determined-ai has an opinion here?

File: examples/diffusion/textual_inversion_stable_diffusion/detsd/pipeline.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (coreystatendet) [reply examples/diffusion/textual_inversion_stable_diffusion/detsd/pipeline.py:78]: Yep same thought re: `train_oncluster`

File: examples/diffusion/textual_inversion_stable_diffusion/detsd/pipeline.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (garrett361) [reply examples/diffusion/textual_inversion_stable_diffusion/detsd/pipeline.py:96]: Refactored as (hopefully) requested.

File: examples/diffusion/textual_inversion_stable_diffusion/detsd/pipeline.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply examples/diffusion/textual_inversion_stable_diffusion/detsd/pipeline.py]: @coreystatendet is right, but it's not forking he's talking about, it's continue training that has that behavior.

@garrett361 the real concern with continuing batch numbers in a new experiment is that every experiment has a searcher, and searchers operate in terms of absolute batches, but they're all hardcoded to start counting at `0` for a particular experiment.  So everything in the system operates like that right now.

Not saying it's ideal how it is... but that's why the system works that way right now.  If somebody were to use the current code with hyperparameter search and they also manually configured a source_checkpoint_trial_id, the resulting hyperparameter search would be broken.  If they never use hpsearch, it's not really a big deal.

[the remark this answers is not in the document]
Comment (garrett361) [reply examples/diffusion/textual_inversion_stable_diffusion/detsd/pipeline.py]: Changed the checkpointing code to follow the docs.

File: examples/diffusion/textual_inversion_stable_diffusion/detsd/pipeline.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (garrett361) [reply examples/diffusion/textual_inversion_stable_diffusion/detsd/pipeline.py:82]: I didn't come up with better names, unfortunately.

File: examples/diffusion/textual_inversion_stable_diffusion/detsd/trainer.py [hunk 10]
```diff
@@ -0,0 +1,690 @@
...
+
+    def _restore_latest_checkpoint(
+        self, core_context: det.core.Context, latest_checkpoint: str
+    ) -> None:
+        """Restores the experiment state to the latest saved checkpoint, if it exists."""
+        with core_context.checkpoint.restore_path(latest_checkpoint) as path:
+            with self.accelerator.local_main_process_first():
+                with open(path.joinpath("metadata.json"), "r") as f:
+                    checkpoint_metadata_dict = json.load(f)
+                    self.steps_completed = checkpoint_metadata_dict["steps_completed"]
```
[the remark this answers is not in the document]
Comment (garrett361) [reply examples/diffusion/textual_inversion_stable_diffusion/detsd/trainer.py]: Rewrote this code to follow the Core API checkpointing docs, too.

File: examples/diffusion/textual_inversion_stable_diffusion/README.md [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+  in
+  the `concepts` section of this advanced config. Use similar
+  filenames and config settings with your own images for a boost in results.
+* There are two forms of regularization which can be used in the `DetSDTextualInversionTrainer` class through its `__init__` args:
+  * `norm_reg_weight`: Penalize the tunable concept embeddings for becoming much larger or smaller than the average SD embedding vector.
+  * `hidden_reg_weight`: Penalize the `encoder_hidden_states` generated by the `text_encoder` from too far from their initialization values.
+
+## Some Tips
+
+* Training images are resized to 512 x 512 pixels by default. Resize your images accordingly will
```
Comment (coreystatendet) [inline examples/diffusion/textual_inversion_stable_diffusion/README.md:190]: typo: "Resize your images" -> "Resizing your images"

**Comment by coreystatendet** [comment]
looks good to me!
